core/src/main/java/com/linecorp/armeria/common/http/HttpMessageAggregator.java

```diff
@@ -34,38 +37,70 @@ protected HttpMessageAggregator(CompletableFuture<AggregatedHttpMessage> future)
         this.future = future;
     }
 
-    protected final CompletableFuture<AggregatedHttpMessage> future() {
-        return future;
-    }
-
     @Override
-    public void onSubscribe(Subscription s) {
+    public final void onSubscribe(Subscription s) {
         subscription = s;
         s.request(Long.MAX_VALUE);
     }
 
-    protected final void add(HttpData data) {
-        final int dataLength = data.length();
-        if (dataLength > 0) {
-            if (contentLength > Integer.MAX_VALUE - dataLength) {
-                clear();
-                subscription.cancel();
-                throw new IllegalStateException("content length greater than Integer.MAX_VALUE");
-            }
+    /**
+     * Handled by {@link #accept(Void, Throwable)} instead,
+     * because this method is not invoked on cancellation and timeout.
+     */
+    @Override
+    public final void onError(Throwable throwable) {}
+
+    /**
+     * Handled by {@link #accept(Void, Throwable)} instead,
+     * because this method is not invoked on cancellation and timeout.
+     */
+    @Override
+    public final void onComplete() {}
 
-            contentList.add(data);
-            contentLength += dataLength;
+    @Override
+    public final void onNext(HttpObject o) {
+        if (o instanceof HttpHeaders) {
+            onHeaders((HttpHeaders) o);
+        } else {
+            onData((HttpData) o);
         }
     }
 
-    protected final void clear() {
-        doClear();
-        contentList.clear();
+    protected abstract void onHeaders(HttpHeaders headers);
+
+    private void onData(HttpData data) {
+        boolean added = false;
+        try {
+            if (future.isDone()) {
+                return;
+            }
+
+            final int dataLength = data.length();
+            if (dataLength > 0) {
+                if (contentLength > Integer.MAX_VALUE - dataLength) {
```

Comment:
How about changing this condition as follows? I think it may be more clear to understand.
``` 
dataLength > Integer.MAX_VALUE - contentLength // the current data size > remaining space
```